### DISCUSSION TITLE
Complementary Filter Configuration

### DISCUSSION THREAD

**Discussion by GNC-JY** [bitcraze/discussions#1001] [Q&A] [answered]
Hello.

I have a question about the Complementary filter that estimates the values of Roll, Pitch, and Yaw.

https://github.com/bitcraze/crazyflie-firmware/blob/cd3f3dfb668261f4461f972be8116d996aa7256e/src/modules/src/sensfusion6.c#L263

Where do the gx, gy, and gz declared here come from? Can I understand this as an accel value?

Also, roll is obtained as atan2, is there a reason why pitch is written asin?

And yaw gets a qz value from qw, is this the value obtained from "https://github.com/bitcraze/crazyflie-firmware/blob/cd3f3dfb668261f4461f972be8116d996aa7256e/src/modules/src/sensfusion6.c#L181C6-L181C6 ?

After changing pitch to atan2, I graphed the estimate value

![image](https://github.com/bitcraze/discussions/assets/110527007/c3d45e64-b934-4fa8-aa63-471f607d3577)

Only the roll moved, but the pitch suddenly divergent at 90 degrees.

Please give me information about the complementary filter configuration.

**Comment by tobbeanton** [comment]
I suggest read [this](https://ahrs.readthedocs.io/en/latest/filters/madgwick.html) documentation for all the detail.

gx, gy, and gz are gyro values, ax, ay and az are accelerometer values.

**Reply by GNC-JY** [reply]
In code, gx, gy, gz is declared of gravX, gravY, gravZ. Is 'grav' means gyro???

https://github.com/bitcraze/crazyflie-firmware/blob/cd3f3dfb668261f4461f972be8116d996aa7256e/src/modules/src/sensfusion6.c#L265

**Reply by knmcguire** [reply]
no that is the gravitational unit vector items. So this is estimated from the accelerometers.

I'm working on a more elaborate explanation on your second part of your question. stay tuned :)

**Comment by knmcguire** [comment, marked as answer]
Hi!

This requires us to dig quite deep in old code so lets see what we can explain and what not.

- The gravx,y,z comes from [this function ](https://github.com/bitcraze/crazyflie-firmware/blob/cd3f3dfb668261f4461f972be8116d996aa7256e/src/modules/src/sensfusion6.c#L91C6-L91C6). @tobbeanton gave another answer with some details

Your questions about why sin for pitch and tan for roll is a bit more difficult to answer as this is based on somebody elses code. So I'm going to try to reverse-engineer it.

- **Euler angle order:** So the function wants to determine the euler angels in the order roll pitch yaw, which means it needs to be calculated in reverse (first yaw, then pitch then roll)
- **Yaw:** Yaw needs to be estimated from previous quarternion estimates.. The accelero meter values do not change when you rotate it around the z axis (since the gravity factor respective from the body fixed axis does not change in orientation)
- **Pitch:** Pitch is a bit easier to calculate. As this is done first we don't have to take the roll factor into account yet: 
![image](https://github.com/bitcraze/discussions/assets/11061260/52a8a735-2c3b-4425-8625-f8ccba74d9a2)
The initial x, y, z in black are the initial coordinate system in similar fashion as global and x', y', z' is the body fixed coordinate system. I am aware that this does not hold with the right hand rule (y should be reverse) but as you got from the note in the code, pitch is in reverse so this is probably fine for now. So red are the gravitational unit factor, of g being the unit factor in global coordinate (always 1) and gz and gy are the gravitational unit factors with rotation of pitch (theta). This is a more 3D view to make it seem easier:
![image](https://github.com/bitcraze/discussions/assets/11061260/d6f94672-537e-445a-913e-549c53c83dc3).
So the reason why it is asin(gx) is because the following. You want to calculate theta with two known values. Here in the corner below you can see that g and gx make a right (corned) triangle, so therefore you can assume that the corner that is part of that triangle is theta as well.
![image](https://github.com/bitcraze/discussions/assets/11061260/f47e70c5-f338-4813-846c-05c258b87402)
So then you can calculate theta as asin(gx/g) and because g is a unit vector (value 1), this simplifies it as theta is asin(gx). But yes you could have also done as atan2(gx/gz), but sin is more of a simplification. 
![image](https://github.com/bitcraze/discussions/assets/11061260/10896fef-6bc5-4e24-ac4c-bd4c1e0fdfa5)
- **Roll:** So now roll will be calculated as last. This is assuming the pitch has already taken place so axis y is the same as y', x'' is the same as the x' as before (rotation around x) ibut different from original x and now and z' is rotated to z'' due to the roll. 
![image](https://github.com/bitcraze/discussions/assets/11061260/5dc2c996-87f1-4d32-b268-6659e24c8942)
This means that the gravitational unit vector of g is now out of plane of x and y... so g now and no longer the unit value of 1 (or at least, pitch value is in there as well now). This can be calculated perhaps but that would would complicate the equation. It was already very difficult to draw in 3d...
![image](https://github.com/bitcraze/discussions/assets/11061260/587f2973-9cd5-4927-91cd-f65cc7de9b0b)
However, gy and gz are still available from the accelerometers and that can be used for calculating the role. 
![image](https://github.com/bitcraze/discussions/assets/11061260/fee55f33-9c6e-4cf8-8c0a-137636c47809)

So that explains at least this part of the code, at least I think also including the conversation to degrees from radians

```
  *yaw = atan2f(2*(qw*qz + qx*qy), qw*qw + qx*qx - qy*qy - qz*qz) * 180 / M_PI_F;
  *pitch = asinf(gx) * 180 / M_PI_F; //Pitch seems to be inverted
  *roll = atan2f(gy, gz) * 180 / M_PI_
```

Quite a lot of assumptions are made here, also that some angles are small enough to disregard, and that the vehicle is in either a still position or moves with a constant velocity. This is where things get weird, since a quadcopter if it tilts, it will **accelerate** in that direction, so it must be detected in the accelerometers as well. However, this filter works great with manual flight... and is better than the Kalman filter which does take that into account... so yeah still a lot of unknowns here and it has been implemented by someone else 10 years ago. 

Interesting fact @krichardsson shared this with me which used to be the calculations before the simplifications but apparently 'didn't work well'

```
  *yaw = atan2(2*q1*q2 - 2*q0*q3, 2*q0*q0 + 2*q1*q1 - 1) * 180 / M_PI;
  *pitch = atan(gx / sqrt(gy*gy + gz*gz)) * 180 / M_PI;
  *roll = atan(gy / sqrt(gx*gx + gz*gz)) * 180 / M_PI;
```
Also the former was prone to divisions by zero (if not using atan2) and also doesn't take the Euler order into account. Perhaps that is why it didn't work.

_Disclaimer: This took a long time to figure out but we were intrigued of why this was :) but don't expect this level of support for further questions . Or else will take forever to help everybody... but it was fun to try to figure this out!_😅

**Reply by GNC-JY** [reply]
Can you tell me which part of 'Didn't work well' didn't work well?

**Reply by knmcguire** [reply]
That's all we know, it just didn't estimate the angles right and it couldn't fly properly? this is still crazyflie 1 era even.

**Reply by GNC-JY** [reply]
No, It fly well. I'm using Crazyflie 2.1. I changed pitch from asin to atan2 (atan2(gx,gz)), and there was no problem with hovering.
However, I am currently working on an algorithm for flight recovery in situations such as rotation for Crazyflie, and I am curious about why the state estimate value of Crazyflie's complementary filter is diverging.

**Reply by GNC-JY** [reply]
I want to make the cause that the complementary filter cannot control the attitude due to 90 degree singularity, but to do so, I need to know exactly how the filter estimates the attitude.

**Reply by knmcguire** [reply]
As I said before in the calculations , atan2 would also work fine for sin, it is just the 'old large formula's' (the last ones) that supposedly didn't work.

Unfortunately, Euler angles are prone to singularities and there is no going around it. If you have a roll of 90 degrees, the Crazyflie shouldn't be able to recover anyway (unless you would like to do flips). If you want to have a singular free filter, it needs to be written fully in quaternions. 

But, what you are seeing in the plot is just a matter of a wrap between -180 degrees and 180 degrees. Perhaps in the estimate you can check when the roll goes close to those values and add them up to each other? Such there is a continuity in there, but I have no idea of the repercussions.

**Reply by GNC-JY** [reply]
Then, in kalman_core.c, what line does the attitude estimator value initialize when the crazyflie is still at a certain angle?

**Reply by knmcguire** [reply]
kalman_core.c is not part of the complimentary filter. If you have a question about the kalman filter you should probably make a new discussion.
